knimin/handlers/barcode_util.py

```diff
@@ -202,12 +462,88 @@ def _build_email(self, login_user, barcode, email_type,
         return subject, body_message
 
 
-def align_with_qiita_categories(samples, categories):
-    # obviously should instead align to the actual sample metadata...
-    aligned = {}
-    for s in samples:
-        aligned[s] = {c: 'LabControl test' for c in categories}
-    return aligned
+def align_with_qiita_categories(samples, categories,
+                                failure_value='pulldown-issue',
+                                omitted_value='Missing: Not provided'):
+    """Obtain sample metadata, and subset to those categories present in Qiita
+
+    Parameters
+    ----------
+    samples : list of str
+        The samples to get metadata for
+    categories : Iterable of str
+        The categories to align against
+    failure_value : str, optional
+        The default value to use for a sample that failed pulldown.
+    omitted_value : str, optional
+        The default value to use for a variable not represented either in Qiita
+        or the extracted metadata.
+
+    Notes
+    -----
+    The env_package variable for failures will be autofilled with "Air" per a
+    request from Gail.
+
+    Any variable in extract metadata that is not represented in Qiita will be
+    silently omitted (e.g., PM_USEFUL).
+
+    Any variable in Qiita that is not represented in the extracted metadata
+    (e.g., qiita_empo_1) will be filled with the omitted_value.
+
+    Returns
+    -------
+    dict of dict
+        A stucture of the metadata per sample. {sample-id: {category: value}}
+    """
+    surveys, failures = db.pulldown(samples)
+
+    # pulldown returns a per-survey (e.g., primary, fermented food, etc) tab
+    # delimited file. What we're doing here is de-serializing those data into
+    # per survey DataFrames, and the concatenating them together such that
+    # each sample ID is a row, each sample ID is only represented once, and the
+    # columns correspond to variables from each survey type.
+    surveys_as_df = []
+    for _, v in sorted(surveys.items()):
+        surveys_as_df.append(pd.read_csv(StringIO.StringIO(v), sep='\t',
+                                         dtype=str).set_index('sample_name'))
+    surveys_as_df = pd.concat(surveys_as_df, axis=1)
+
+    # columns in Qiita are lower case
+    surveys_as_df.columns = [c.lower() for c in surveys_as_df.columns]
+
+    # subset the frame to the overlapping columns
+    categories = set(categories)
+    column_overlap = [c for c in surveys_as_df.columns if c in categories]
```

Comment:
Index objects support some set operations so this could be: `surveys_as_df.columns.intersection(categories)`.